odoo/sql_db.py

```diff
@@ -510,32 +575,46 @@ class TestCursor(BaseCursor):
                                     |
               cr.execute(query)     | query
                                     |
-              cr.commit()           | SAVEPOINT test_cursor_N
+              cr.commit()           | RELEASE SAVEPOINT test_cursor_N
+                                    | SAVEPOINT test_cursor_N (lazy)
                                     |
-              cr.rollback()         | ROLLBACK TO SAVEPOINT test_cursor_N
+              cr.rollback()         | ROLLBACK TO SAVEPOINT test_cursor_N (if savepoint)
                                     |
-              cr.close()            | ROLLBACK TO SAVEPOINT test_cursor_N
-                                    |
-
+              cr.close()            | ROLLBACK TO SAVEPOINT test_cursor_N (if savepoint)
+                                    | RELEASE SAVEPOINT test_cursor_N (if savepoint)
     """
-    _savepoint_seq = itertools.count()
-
+    _cursors = []
     def __init__(self, cursor, lock):
         super().__init__()
         self._closed = False
         self._cursor = cursor
         # we use a lock to serialize concurrent requests
         self._lock = lock
         self._lock.acquire()
+        self._cursors.append(self)
         # in order to simulate commit and rollback, the cursor maintains a
-        # savepoint at its last commit
-        self._savepoint = "test_cursor_%s" % next(self._savepoint_seq)
-        self._cursor.execute('SAVEPOINT "%s"' % self._savepoint)
+        # savepoint at its last commit, the savepoint is created lazily
+        self._savepoint = self._cursor.savepoint(flush=False)
+
+    @check
+    def execute(self, *args, **kwargs):
+        if not self._savepoint:
+            self._savepoint = self._cursor.savepoint(flush=False)
+
+        return self._cursor.execute(*args, **kwargs)
 
     def close(self):
         if not self._closed:
             self.rollback()
             self._closed = True
+            if self._savepoint:
+                self._savepoint.close(rollback=False)
+            if self._cursors[-1] is not self:
+                _logger.warning("Found different un-closed cursor when trying to close %s: %s",
+                                self, self._cursors)
+                while self._cursors[-1] is not self:
+                    self._cursors.pop()
+            self._cursors.pop()
```

Comment:
When interleaving test cursors, the first one to be closed logs a warning, while the second one crashes because popping from an empty `self._cursor`!  Maybe you should write a test case for this one...

Reply:
Hmmm that's a good point, things also get weird if we never find a cursor in the stack for some reason, but it should probably not crash either.

Reply:
Well turns out it doesn't actually break, because the `rollback` will break when it tries to release the savepoint (which already doesn't exist anymore).